ArdusatSDK.cpp

```diff
@@ -453,266 +217,1101 @@ const char * valuesToCSV(const char *sensorName, unsigned long timestamp, int nu
+Magnetic::Magnetic(void) :
+  gaussScale(LSM303_MAG_SCALE4GAUSS)
 {
-  _resetOutBuf();
-  _writeJSONValue(_getOutBuf(), sensor_name, unit_to_str(temp.header.unit), temp.t);
-  return _getOutBuf();
+  this->initializeHeader(SENSORID_ADAFRUIT9DOFIMU, DATA_UNIT_MICROTESLA, magnetic_sensor_name);
 }
 
-const char * luminosityToJSON(const char *sensor_name, luminosity_t & lum)
+/**
+ * @brief   Sets the gauss scale configuration variable
+ * @ingroup magnetic
+ *
+ * @param gaussScale Advanced configuration for magnetometer's scale
+ *     - LSM303_MAG_SCALE1_3GAUSS
+ *     - LSM303_MAG_SCALE2GAUSS
+ *     - LSM303_MAG_SCALE2_5GAUSS
+ *     - LSM303_MAG_SCALE4GAUSS (Default)
+ *     - LSM303_MAG_SCALE4_7GAUSS
+ *     - LSM303_MAG_SCALE5_6GAUSS
+ *     - LSM303_MAG_SCALE8GAUSS
+ *     - LSM303_MAG_SCALE12GAUSS
+ */
+void Magnetic::configure(lsm303_mag_scale_e gaussScale) {
+  this->gaussScale = gaussScale;
+  this->initialized = false;
+}
+
+/**
+ * @brief   Initializes the sensor with any set configurations or defaults
+ * @ingroup magnetic
+ *
+ * @retval true  Successfully initialized
+ * @retval false Failed to initialize
+ */
+boolean Magnetic::initialize(void) {
+  return lsm303_mag_init(this->gaussScale);
+}
+
+/**
+ * @brief   Takes a reading from the sensor
+ * @ingroup magnetic
+ */
+void Magnetic::read(void) {
+  this->header.timestamp = millis();
+  lsm303_getMag(&(this->x), &(this->y), &(this->z));
+}
+
+/**
+ * @brief   Returns last read value in CSV format
+ * @ingroup magnetic
+ * @param   sensorName The text to display next to the value
+ * @return  sensor readings in CSV format or empty string if uninitialized
+ */
+const char * Magnetic::toCSV(const char * sensorName) {
+  if (this->initialized) {
+    return valuesToCSV(sensorName, this->header.timestamp, 3,
+                       this->x, this->y, this->z);
+  } else {
+    return "";
+  }
+}
+
+/**
+ * @brief   Returns last read value in JSON format
+ * @ingroup magnetic
+ * @param   sensorName The text to display next to the value
+ * @return  sensor readings in JSON format or empty string if uninitialized
+ */
+const char * Magnetic::toJSON(const char * sensorName) {
+  if (this->initialized) {
+    return valuesToJSON(sensorName, this->header.unit, 3, "X", this->x,
+                              "Y", this->y, "Z", this->z);
+  } else {
+    return "";
+  }
+}
+
+
+/**************************************************************************//**
+ * @brief   Constructs Orientation calculation object
+ * @ingroup orientation
+ *
+ * Example Usage:
+ * @code
+ *     Orientation orient;                               // Instantiate sensor object
+ *     orient.begin();                                   // Initialize sensor
+ *     Serial.println(orient.readToJSON("orientation")); // Read and print values in JSON
+ * @endcode
+ *****************************************************************************/
+Orientation::Orientation(void) {
+  Acceleration accel;
+  Magnetic mag;
```

Comment:
This doesn't really work as written. This class stores pointers to objects that were just instantiated here on the stack, so they go out of scope and are destructed as soon as this function body finishes. That leaves you with dangling pointers `accel` and `mag`, which could do anything when you dereference them...

You'll need to heap-allocate these objects (e.g. use `new`) to get them to persist, but then you've opened yourself up to a memory leak; they won't be removed by the default destructor since you're just storing raw pointers to them. You'll have to have an internal variable that keeps track of allocation (e.g. whether this class allocated your pointer references) and manually call `delete` on the pointers if so in a custom destructor.